test/common/buffer/buffer_speed_test.cc

```diff
@@ -0,0 +1,221 @@
+#include "common/common/assert.h"
+
+#include "absl/strings/string_view.h"
+#include "testing/base/public/benchmark.h"
+
+namespace Envoy {
+
+// Test the creation of an empty OwnedImpl.
+static void BufferCreateEmpty(benchmark::State& state) {
+  uint64_t length = 0;
+  for (auto _ : state) {
+    Buffer::OwnedImpl buffer;
+    length += buffer.length();
+  }
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(BufferCreateEmpty);
+
+// Test the creation of an OwnedImpl with varying amounts of content.
+static void BufferCreate(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  const absl::string_view input(data);
+  uint64_t length = 0;
+  for (auto _ : state) {
+    Buffer::OwnedImpl buffer(input);
+    length += buffer.length();
+  }
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(BufferCreate)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+// Test the appending of varying amounts of content from a string to an OwnedImpl.
+static void BufferAddString(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  const absl::string_view input(data);
+  Buffer::OwnedImpl buffer(input);
+  for (auto _ : state) {
+    buffer.add(data);
+    // Drain from the front of the Buffer to keep the test's memory usage fixed.
+    // @note Ideally we could use state.PauseTiming()/ResumeTiming() to exclude
+    // the time spent in the drain operation, but those functions themselves are
+    // heavyweight enough to cloud the measurements:
+    // https://github.com/google/benchmark/issues/179
+    buffer.drain(input.size());
+  }
+  uint64_t length = buffer.length();
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(BufferAddString)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+// Variant of BufferAddString that appends from another Buffer::Instance
+// rather than from a string.
+static void BufferAddBuffer(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  const absl::string_view input(data);
+  const Buffer::OwnedImpl to_add(data);
+  Buffer::OwnedImpl buffer(input);
+  for (auto _ : state) {
+    buffer.add(to_add);
+    // Drain from the front of the Buffer to keep the test's memory usage fixed.
+    buffer.drain(input.size());
+  }
+  uint64_t length = buffer.length();
+  benchmark::DoNotOptimize(length);
+}
+BENCHMARK(BufferAddBuffer)->Arg(1)->Arg(4096)->Arg(16384)->Arg(65536);
+
+static void BufferDrain(benchmark::State& state) {
+  const std::string data(state.range(0), 'a');
+  const absl::string_view input(data);
+  const Buffer::OwnedImpl to_add(data);
+  Buffer::OwnedImpl buffer(input);
+
+  // On each iteration of the benchmark, add N bytes and drain a multiple of N, as specified
+  // by DrainCycleRatios. This exercises full-slice, partial-slice, and multi-slice code paths
+  // in the Buffer's drain implementation.
+  constexpr size_t DrainCycleSize = 7;
+  constexpr double DrainCycleRatios[DrainCycleSize] = {0.0, 1.5, 1, 1.5, 0, 2.0, 1.0};
```

Comment:
I'd definitely like to see trickle-add and trickle-drain tests added - add one byte at a time many many times, remove one byte at a time many many times.    Should be easy enough and while I hope we never have a buffer impl that sucks for those trivial use cases I think it's worth covering :-)

Reply:
Sounds good to me. Currently, the `BufferMovePartial` test exercises the moving of a byte at a time from one buffer to another, which happens to exercise the add/drain code paths today but might not in the future (if we ever add reference-copying to do zero-copy moves between buffers, for example). So I'll add benchmark tests that directly exercise 1-byte adds and drains.